tools/azure-sdk-tools/devtools_testutils/azure_recorded_testcase.py

```diff
@@ -261,3 +189,105 @@ def generate_sas(self, *args, **kwargs):
+        if use_cli.lower() == "true":
+            _LOGGER.info("Environment variable AZURE_TEST_USE_CLI_AUTH set to 'true'. Using AzureCliCredential.")
+            from azure.identity import AzureCliCredential
+
+            if is_async:
+                from azure.identity.aio import AzureCliCredential
+            return AzureCliCredential()
+        # User-based authentication through Visual Studio Code, if requested
+        if use_vscode.lower() == "true":
+            _LOGGER.info(
+                "Environment variable AZURE_TEST_USE_VSCODE_AUTH set to 'true'. Using VisualStudioCodeCredential."
+            )
+            from azure.identity import VisualStudioCodeCredential
+
+            if is_async:
+                from azure.identity.aio import VisualStudioCodeCredential
+            return VisualStudioCodeCredential()
+        # User-based authentication through Azure Developer CLI (azd), if requested
+        if use_azd.lower() == "true":
+            _LOGGER.info(
+                "Environment variable AZURE_TEST_USE_AZD_AUTH set to 'true'. Using AzureDeveloperCliCredential."
+            )
+            from azure.identity import AzureDeveloperCliCredential
+
+            if is_async:
+                from azure.identity.aio import AzureDeveloperCliCredential
+            return AzureDeveloperCliCredential()
+
+        # Service principal authentication
+        if tenant_id and client_id and secret:
+            _LOGGER.info(
+                "Service principal client ID, secret, and tenant ID detected. Using ClientSecretCredential.\n"
+                "For user-based auth, set AZURE_TEST_USE_PWSH_AUTH or AZURE_TEST_USE_CLI_AUTH to 'true'."
+            )
+            from azure.identity import ClientSecretCredential
+
+            if is_async:
+                from azure.identity.aio import ClientSecretCredential
+            return ClientSecretCredential(tenant_id=tenant_id, client_id=client_id, client_secret=secret)
+
+        # If AzurePipelinesCredential is detected, use it.
+        service_connection_id = os.environ.get("AZURESUBSCRIPTION_SERVICE_CONNECTION_ID")
+        client_id = os.environ.get("AZURESUBSCRIPTION_CLIENT_ID")
+        tenant_id = os.environ.get("AZURESUBSCRIPTION_TENANT_ID")
+        system_access_token = os.environ.get("SYSTEM_ACCESSTOKEN")
+        if service_connection_id and client_id and tenant_id and system_access_token:
+            from azure.identity import AzurePipelinesCredential
+            if is_async:
+                from azure.identity.aio import AzurePipelinesCredential
+            return AzurePipelinesCredential(
+                tenant_id=tenant_id,
+                client_id=client_id,
+                service_connection_id=service_connection_id,
+                system_access_token=system_access_token)
+        # This is for testing purposes only, to ensure that the AzurePipelinesCredential is used when available
```

Comment:
Should we remove this commented section now that things have been tested?

Reply:
We are still validating if AzurePipelinesCredential can solve the performance issue.

Let's keep it for a while in case we want to use APC for sure.